### PR TITLE
url: catch Failure & propose a `parse_opt`

### PR DESCRIPTION
**Pull request by rjbou**
instead of internal error

### CODE REVIEW

**Comment by AltGr** [comment]
I am not sure it is a good idea to hard fail from this low-level module: couldn't that happen when parsing URLs from files, e.g. from a custom repository (Whereas the `Failure` would have been caught in `OpamFile`) ?

**Comment by rjbou** [comment]
Yup, maybe add an `OpamUrl.parse_opt` and treat the `None ` case for each call (instead of treating exception handling)? It requests several modifications (around 30 places), but at least failures would be contained (and treated!).

File: src/client/opamAdminRepoUpgrade.ml [hunk 1]
```diff
@@ -177,7 +177,10 @@ let do_upgrade repo_root =
     let url_md5 = Hashtbl.create 187 in
     let () =
       OpamFile.Lines.read_opt cache_file +! [] |> List.iter @@ function
-      | [url; md5] -> Hashtbl.add url_md5 (OpamUrl.of_string url) (OpamHash.of_string md5)
+      | [url; md5] ->
+        OpamStd.Option.iter
+          (fun url -> Hashtbl.add url_md5 url (OpamHash.of_string md5))
+          (OpamUrl.parse_opt ~handle_suffix:false url)
```
[the remark this answers is not in the document]
Comment (rjbou) [reply src/client/opamAdminRepoUpgrade.ml]: The next file is an indicator that cache is corrupted, while we shouldn't find a non available url in this file. This said, we can ignore some (malfomed) urls in this cache, as it is used to get hashes, and [url](https://github.com/ocaml/opam/pull/3946/files/53d54e1ecce72d698d3cc56b0461d301fa054563#diff-2053ac7e88a141d8e27dd39483a1b885R246) given in argument is already checked via the [comp](https://github.com/ocaml/opam/pull/3946/files/53d54e1ecce72d698d3cc56b0461d301fa054563#diff-2053ac7e88a141d8e27dd39483a1b885R216) file parsing. Plus, if the url is missing from cache (malformed in cache, but good in comp file), there is a [fallback](https://github.com/ocaml/opam/pull/3946/files/53d54e1ecce72d698d3cc56b0461d301fa054563#diff-2053ac7e88a141d8e27dd39483a1b885R188) to retrieve the good hash.

File: src/client/opamCommands.ml [hunk 2]
```diff
@@ -1793,7 +1792,12 @@ let repository =
           with Failure _ -> failwith ("Invalid quorum: "^quorum)
       in
       OpamRepositoryState.with_ `Lock_write gt (fun rt ->
-          let rt = OpamRepositoryCommand.add rt name url trust_anchors in
+          let rt =
+            match OpamUrl.parse_opt ?backend url with
+            | Some url ->
+              OpamRepositoryCommand.add rt name url trust_anchors
+            | None -> rt
```
[the remark this answers is not in the document]
Comment (rjbou) [reply src/client/opamCommands.ml]: here we must fail!

**Pull request merged**